### PR TITLE
Move conflicting dependencies into PubGrub

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This revives a PR from long ago (https://github.com/astral-sh/uv/pull/383 and https://github.com/zanieb/pubgrub/pull/24) that modifies how we deal with dependencies that are declared multiple times within a single package.

To quote from the originating PR:

> Uses an experimental pubgrub branch (#370) that allows us to handle multiple version ranges for a single dependency to the solver which results in better error messages because the derivation tree contains all of the relevant versions. Previously, the version ranges were merged (by us) in the resolver before handing them to pubgrub since only one range could be provided per package. Since we don't merge the versions anymore, we no longer give the solver an empty range for conflicting requirements; instead the solver comes to that conclusion from the provided versions. You can see the improved error message for direct dependencies in [this snapshot](https://github.com/astral-sh/puffin/pull/383/files#diff-a0437f2c20cde5e2f15199a3bf81a102b92580063268417847ec9c793a115bd0).

The main issue with that PR was around its handling of URL dependencies, so this PR _also_ refactors how we handle those. Previously, we stored URL dependencies on `PubGrubPackage`, but they were omitted from the hash and equality implementations of `PubGrubPackage`. This led to some really careful codepaths wherein we had to ensure that we always visited URLs before non-URL packages, so that the URL-inclusive versions were included in any hashmaps, etc. I considered preserving this approach, but it would require us to rely on lots of internal details of PubGrub (since we'd now be relying on PubGrub to merge those packages in the "right" order).

So, instead, we now _always_ set the URL on a given package, whenever that package was _given_ a URL upfront. I think this is easier to reason about: if the user provided a URL for `flask`, then we should just always add the URL for `flask`. If we see some other URL for `flask`, we error, like before. If we see some unknown URL for `flask`, we error, like before.

Closes https://github.com/astral-sh/uv/issues/1522.

Closes https://github.com/astral-sh/uv/issues/1821.

Closes https://github.com/astral-sh/uv/issues/1615.

### CODE REVIEW

**Comment by charliermarsh** [comment]
(No reviews yet, this doesn't pass the test suite.)

**Comment by charliermarsh** [comment]
\cc @zanieb for visibility.

File: crates/uv/tests/pip_install_scenarios.rs [hunk 1]
```diff
@@ -1940,10 +1941,10 @@ fn transitive_prerelease_and_stable_dependency_many_versions() {
...
-          And because only crow<2.0.0b1 is available, we can conclude that all versions of bluebird depend on crow<2.0.0b1.
-          And because albatross==1.0.0 depends on crow>=2.0.0b1 and only albatross==1.0.0 is available, we can conclude that all versions of bluebird and all versions of albatross are incompatible.
-          And because you require bluebird and you require albatross, we can conclude that the requirements are unsatisfiable.
+      ╰─▶ Because only albatross==1.0.0 is available and albatross==1.0.0 depends on crow>=2.0.0b1, we can conclude that all versions of albatross depend on crow>=2.0.0b1.
+          And because only crow<2.0.0b1 is available, we can conclude that all versions of albatross depend on crow>3.0.0.
+          And because bluebird==1.0.0 depends on crow and only bluebird==1.0.0 is available, we can conclude that all versions of albatross and all versions of bluebird are incompatible.
+          And because you require albatross and you require bluebird, we can conclude that the requirements are unsatisfiable.
```
Comment (charliermarsh) [inline crates/uv/tests/pip_install_scenarios.rs:1947]: I believe these are generally the result of dependencies being presented to PubGrub in a different order.

File: crates/uv/tests/pip_compile.rs [hunk 2]
```diff
@@ -1272,15 +1272,13 @@ fn conflicting_repeated_url_dependency_version_mismatch() -> Result<()> {
...
 
     ----- stderr -----
-      × No solution found when resolving dependencies:
-      ╰─▶ your requirements cannot be used because there are conflicting URLs for
-          package `werkzeug`:
-          - https://files.pythonhosted.org/packages/bd/24/11c3ea5a7e866bf2d97f0501d0b4b1c9bbeade102bb4b588f0d2919a5212/Werkzeug-2.0.1-py3-none-any.whl
-          - https://files.pythonhosted.org/packages/ff/1d/960bb4017c68674a1cb099534840f18d3def3ce44aed12b5ed8b78e0153e/Werkzeug-2.0.0-py3-none-any.whl
+    error: There are conflicting URLs for package `werkzeug`:
+    - https://files.pythonhosted.org/packages/bd/24/11c3ea5a7e866bf2d97f0501d0b4b1c9bbeade102bb4b588f0d2919a5212/Werkzeug-2.0.1-py3-none-any.whl
+    - https://files.pythonhosted.org/packages/ff/1d/960bb4017c68674a1cb099534840f18d3def3ce44aed12b5ed8b78e0153e/Werkzeug-2.0.0-py3-none-any.whl
```
Comment (zanieb) [inline crates/uv/tests/pip_compile.rs:1281]: Is this an improvement? Should we clarify that the URLs come from the provided requirements?

File: crates/uv/tests/pip_compile.rs [hunk 3]
```diff
@@ -1598,8 +1594,8 @@ dependencies = ["django==5.0b1", "django==5.0a1"]
 
     ----- stderr -----
       × No solution found when resolving dependencies:
-      ╰─▶ my-project cannot be used because there are conflicting versions for
-          `django`: `django==5.0b1` does not intersect with `django==5.0a1`
+      ╰─▶ Because my-project depends on django==5.0b1 and my-project depends on
+          django==5.0a1, we can conclude that the requirements are unsatisfiable.
```
Comment (zanieb) [inline crates/uv/tests/pip_compile.rs:1598]: Nice. It'll be better once we combine clauses #1009, maybe someday we should even track the source of dependencies so we can say what file each of these came from.

File: crates/uv/tests/pip_compile.rs [hunk 4]
```diff
@@ -3754,6 +3750,40 @@ fn compile_constraints_compatible_url() -> Result<()> {
     Ok(())
 }
 
+/// Resolve a package from a `requirements.in` file, with a `constraints.txt` file pinning it to
+/// a specific URL.
```
Comment (zanieb) [inline crates/uv/tests/pip_compile.rs:3754]: This reads a little confusing

File: crates/uv/tests/pip_compile.rs [hunk 5]
```diff
@@ -3910,3 +3940,68 @@ fn no_deps_invalid_extra() -> Result<()> {
...
+            .arg("--constraint")
+            .arg("constraints.txt"), @r###"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+
+    ----- stderr -----
+      × No solution found when resolving dependencies:
+      ╰─▶ Because you require filelock==1.0.0 and you require filelock==3.8.0, we
+          can conclude that the requirements are unsatisfiable.
```
Comment (zanieb) [inline crates/uv/tests/pip_compile.rs:4002]: This one definitely makes me want to open an issue to display the locations of direct requirements in error messages.

File: crates/uv/tests/pip_install_scenarios.rs (hunk 1 above)
Comment (zanieb) [reply crates/uv/tests/pip_install_scenarios.rs:1947]: Yeah it's a little weird but a known issue.

File: Cargo.toml [hunk 6]
```diff
@@ -66,7 +66,7 @@ owo-colors = { version = "4.0.0" }
 petgraph = { version = "0.6.4" }
 platform-info = { version = "2.0.2" }
 plist = { version = "1.6.0" }
-pubgrub = { git = "https://github.com/zanieb/pubgrub", rev = "9b6d89cb8a0c7902815c8b2ae99106ba322ffb14" }
+pubgrub = { git = "https://github.com/zanieb/pubgrub", rev = "5d0270cc2932ab233bb7f66a09fb0764fed7154e" }
```
Comment (zanieb) [inline Cargo.toml:69]: Let's make sure to merge upstream and use the `main` commit before merging this one.

**Review by zanieb — APPROVED** [review]
Looks nice. Is there anything you wanted me to give a closer look?

**Pull request merged**